[ Feature ] Avatar(view/edit/AvatarList) 컴포넌트 제작

SeungHyune opened the pull request:
## 📝 설명
Avatar 컴포넌트, AvatarList 컴포넌트 제작했습니다.

[스토리북 링크](https://6633bd85fb75146bc1604a73-xjskmugxoi.chromatic.com/?path=/docs/components-avatar--docs)
<!-- PR에 대한 설명입니다. -->

## ✅ PR 유형

<!--
	팀원이 어떤 작업을 하였는지 쉽게 알아볼 수 있게 도와주는 부분입니다.
-->

- [x] 새로운 기능 추가
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드 리팩토링
- [ ] 파일 혹은 폴더명 수정 / 삭제
- [ ] 기타

## 💻 작업 내용
### Avatar 컴포넌트 제작 - view, edit type으로 기능을 분리했습니다.
  - **view :** `userId` 값을 전달하면 해당 유저의 정보로 이동되도록 구현했습니다. (만약 userId 값을 전달하지 않으면 링크 이동은 제한됩니다.)
  - **edit :** type을 edit으로 지정하면 `<input type="file" />`이 생성되며 Avatar를 클릭 시 input을 클릭하게 됩니다. 이후 로직은 Avatar을 변경 가능하게 되고 변경된 데이터는 File로 onChangeAvatar 함수로 전달받을 수 있습니다.

나머지 타입들은 프로젝트에서 사용되는 사이즈를 임의로 저장하여 정해진 사이즈만 사용하도록 작업했습니다.
`avatarImage`를 통해 초기 데이터가 있는 경우 해당 데이터 이미지가 보이도록 설정했습니다. 만약 `null`이라면 기본 이미지가 보이도록 설정했습니다.

type이 edit인데 `onChangeAvatar` 함수를 props로 전달하지 않은 경우에는 변경한 데이터를 외부에서 전달받을 수 없으므로 input 자체 클릭을 막았습니다.

### AvatarList 컴포넌트 제작
- avatars 데이터가 0개인 경우 아무것도 보여주지 않습니다.
- avatars 데이터가 1개인 경우 Avatar 컴포넌트를 보여줍니다.
- avatars 데이터가 2개 이상인 경우 앞에서부터 2개의 데이터 이미지를 보여주고 해당 AvatarList를 클릭 시 모든 avatars 데이터를 보이도록 작업했습니다. id 값이 있는 경우 클릭 시 이동도 가능합니다.

<!-- PR 본문을 입력하세요. -->

## 💬 PR 포인트
`Avatar` 컴포넌트에 type을 통해 두 가지 일을 하는 컴포넌트를 만들었는데, 이렇게 하는 것이 좋은 건지 잘 모르겠습니다. **컴포넌트는 하나의 일만 하는 게 좋다**고 생각해서  `<AvatarView>, `<AvatarEdit>` 이렇게 나누는 것이 더 좋을 것 같다는 생각도 드네요.

추가로 `AvatarList`를 클릭했을 때 동작에 대한 고민이 많았습니다. 어떻게 처리하는 것이 좋을지 좋은 아이디어 있으시면 말씀 부탁드리겠습니다.
<!-- PR 리뷰 시 공유 사항 또는 유심히 보면 좋을 부분을 설명합니다. -->

SeungHyune committed:
💎feat: useClickAvatar hook 추가
- type이 view인 경우 click을 통한 user 페이지로 이동이 가능합니다.
- type이 edit인 경우 click을 통한 input(file) 요소를 클릭합니다.

SeungHyune committed:
💎feat: useChangeImage hook 추가
- type edit인 경우 input을 클릭 후 이미지를 변경하면 이미지가 변경되고 변경된 이미지를 onChangeAvatar 함수의 file로 전달합니다.

SeungHyune committed:
📝docs: 첨부 파일 확장자 및 avatar size에 대한 상수 파일 추가

SeungHyune committed:
📝docs: avatarImage가 없는 경우(null인 경우)의 기본 이미지 추가

SeungHyune committed:
🪄change: Avatar Hooks 파일 index.ts로 관리

SeungHyune committed:
📝docs: Delete `␍` eslint/prettier 오류 해결

SeungHyune committed:
💎feat: Avatar 컴포넌트 추가
- type을 받아 view, edit 역할을 하는 컴포넌트
- view 아바타를 볼 수 있고 아바타를 클릭 시 해당 유저 페이지로 이동합니다.
- edit 아바타를 볼 수 있고 아바타를 클릭 시 아바타 이미지를 변경할 수 있습니다.

SeungHyune committed:
💎feat: Avatar Storybook 추가

SeungHyune committed:
📝docs: change 이모지가 안 보여서 새로운 이모지로 변경

SeungHyune committed:
🎨style: Avatar Style 수정

SeungHyune committed:
🎨style: Avatar style 수정 및 storybook 수정(default data 추가)

SeungHyune committed:
🎨style: Avatar 컴포넌트 style 수정

SeungHyune committed:
💎feat: useClickAvatarList hook 추가
- toggle 상태를 관리하는 hook
- AvatarList를 클릭하면 현재 toggle의 반대 boolean 값으로 변경된다.
- AvatarList가 아닌 배경을 클릭하면 toggle이 false 값으로 변경된다.

SeungHyune committed:
💎feat: AvatarList 컴포넌트 추가
- AvatarList를 하나의 Avatar로 보이도록 하는 컴포넌트입니다.
- 클릭 시 AvatarList의 목록을 모두 확인할 수 있고, 이동 또한 가능합니다.

SeungHyune committed:
💎feat: AvatarList가 2개 미만인 경우에 대한 예외 처리 추가

SeungHyune committed:
🎨style: Avatar style 수정
- Avatar 이미지가 콘텐츠 박스를 넘치는 부분 수정

SeungHyune committed:
💎feat: AvatarList Storybook 추가

SeungHyune committed:
🚜change: Avatar Storybook 수정
- default prop 데이터 수정

Jae0o commented on src/app/_components/client/AvatarList/AvatarList.tsx:
```diff
@@ -0,0 +1,96 @@
+        <div
+          className="fixed size-[100%] inset-0 bg-black/50 z-0 cursor-pointer"
+          onClick={handleClickBackground}
+        ></div>
```
[ P3 ]

내부에 들어갈 요소가없다면 아래와같이 수정하면 좋을것같습니다.

```suggestion
        />
```

Jae0o commented on src/app/_components/client/AvatarList/AvatarList.tsx:
```diff
@@ -0,0 +1,96 @@
+          ))}
+        </div>
+
+        <motion.ul
+          className="absolute left-0 top-0 flex z-10"
+          initial={{ opacity: 0, x: "-100%" }}
+          animate={toggle ? "open" : "closed"}
+          variants={variants}
+          whileTap={{ scale: 0.9 }}
+        >
+          {avatars.map(({ id, avatarImage }) => (
+            <li
+              className="mr-2 rounded-full overflow-hidden"
+              key={id}
+            >
+              <Avatar
+                avatarImage={avatarImage}
+                userId={id}
+              />
+            </li>
+          ))}
+        </motion.ul>
```
[ P2 ]

최적화를 위해 `framer-motion`에서 지원하는 `AnimatePrecense` 기능을 활용해 `exit` animation 을 구현해보는건 어떨까요?

현재는 ul태그가 완전히 unmount가 되는게아닌 트릭과같이 숨겨지는 방식이다 보니 필요할때만 mount시켜줄 수 있는 방법이 좋을것 같다고 생각이 들었어요!

```suggestion
        <AnimatePresence>
          {toggle && (
            <motion.ul
              className="absolute left-0 top-0 flex z-10"
              initial={{ opacity: 0, x: "-100%" }}
              animate={{ opacity: 1, x: 0 }}
              exit={{ opacity: 0, x: "-100%" }}
              whileTap={{ scale: 0.9 }}
            >
              {avatars.map(({ id, avatarImage }) => (
                <li
                  className="mr-2 rounded-full overflow-hidden"
                  key={id}
                >
                  <Avatar
                    avatarImage={avatarImage}
                    userId={id}
                  />
                </li>
              ))}
            </motion.ul>
          )}
        </AnimatePresence>
```

SeungHyune replied:
`AnimatePrecense`에 대해서는 모르고 있었는데 좋은 내용 공유해 주셔서 감사합니다 !! 👏
좋은 제안인 것 같습니다 수정해놓겠습니다 ㅎㅎ 😊

Jae0o commented on src/app/_components/client/AvatarList/hooks/useClickAvatarList/useClickAvatarList.tsx:
[ P2 ] 

해당 hook 내부 로직의 성향이 toggle과 동일한것같아서 useToggle이라는 공통 hook으로 빼는건어떨까요?

그럼 자유롭게 사용할 수 있을것같아요!
그리고 공용 hook인 만큼 내부에 단순 toggle 목적의 메서드와 추가 메서드도 제공되면 좋을것같아요!
( 네이밍은 제가 생각나는 임시 네이밍일뿐 편하게 작업하셔도될것같습니다. )

- handleToggle : state 단순 toggle해주는 메서드
- handleToggleOff : `false`
- handleToggleOn : `true`

SeungHyune replied:
맞습니다 ! 이번에 `Modal` 컴포넌트 만들 때도 동일한 hook을 만들 게 되더라고요 ! 👍👍
유용하게 사용할 수 있을 것 같아 공용 hook으로 하나 만들어 놓겠습니다 😊

Jae0o commented on src/app/_components/client/AvatarList/AvatarList.tsx:
```diff
@@ -0,0 +1,96 @@
+          {avatars.slice(0, 2).map(({ id, name, avatarImage }, index) => (
+            <div
+              key={id}
+              className={`relative w-[3rem] h-[3rem] rounded-full overflow-hidden ${index === 1 && "-ml-[2.2rem]"} ${index === 1 && "mt-[1rem]"} ${index === 0 && "z-10"}`}
```
[ P4 ]

twJoin을 통해 조건부 처리를 해서 하나로 묶어주는건 어떨까요?

```suggestion
              className={twJoin(
                "relative w-[3rem] h-[3rem] rounded-full overflow-hidden",
                index === 1 && "ml-[2.2rem] mt-[1rem]",
                index === 0 && "z-10"
              )}
```

SeungHyune replied:
넵 ~ 수정해놓겠습니다 !! 😀

Jae0o commented on src/app/_components/client/AvatarList/AvatarList.tsx:
```diff
@@ -0,0 +1,96 @@
+"use client";
+
+import profileImage from "../../../../../public/assets/userProfile.svg";
```
[ P2 ]

디자이너분과 상의후 해당 asset을 Icon으로 만들어도 좋을것 같습니다!

SeungHyune replied:
그럼 일단 SVG 아이콘으로 만들어 놓고 추후 디자인이 나온다면 그때 수정해놓겠습니다!

Jae0o commented on src/app/_components/client/Avatar/Avatar.tsx:
```diff
@@ -0,0 +1,64 @@
+      onClick={handleClickAvatar}
+    >
+      <div className="size-[100%] bg-gray-300 rounded-circle  overflow-hidden">
+        <Image
+          className="object-cover rounded-circle"
+          src={preview || profileImage}
+          alt="프로필 이미지 사진"
+          sizes="100%"
+          fill
+          priority
+        />
+      </div>
+
+      {type === "edit" && (
+        <>
+          <span
+            className={`absolute bottom-0 right-0 ${avatarIconSize} bg-[#F1F1F1] rounded-full`}
+          ></span>
+          <input
+            className="hidden"
+            ref={inputRef}
+            type="file"
+            name="profile"
+            id="profile"
+            onChange={handleChangeImage}
+          />
+        </>
+      )}
```
[ P2 ]

해당 Avatar Component가 특히 `RSC`로 제작되면 좋을것 같다고 생각이 들었습니다!

현재 Edit Type때문에 RSC가 아닌 RCC로 제작되어지는것같은데 아래 처럼 Image 태그와 Input을 별도의 하나의 `RCC`로 제작하고 해당 Avatar 의 기본은 `RSC`로 가져가는건 어떨까요?

```suggestion
        {type === "view" && (
          <Image
            className="object-cover rounded-circle"
            src={preview || profileImage}
            alt="프로필 이미지 사진"
            sizes="100%"
            fill
            priority
          />
        )}
      </div>
      
      
// --------- 아래 부분을 하나의 Component로 분리! 하면 어떨까요?


      {type === "edit" && (
        <>
          <Image
            className="object-cover rounded-circle"
            src={preview || profileImage}
            alt="프로필 이미지 사진"
            sizes="100%"
            fill
            priority
          />
   
          <span
            className={`absolute bottom-0 right-0 ${avatarIconSize} bg-[#F1F1F1] rounded-full`}
          ></span>
          <input
            className="hidden"
            ref={inputRef}
            type="file"
            name="profile"
            id="profile"
            onChange={handleChangeImage}
          />
        </>
      )}
```

SeungHyune replied:
저도 현재와 같이 `Avatar` 컴포넌트 내부에서 `type`을 받아 `view, edit`을 분리하는 로직이라면 각각을 컴포넌트로 분리하는 것이 좋을 것 같습니다 !!

수정하겠습니다 ! 😊

SeungHyune replied:
코드를 수정하다가 문득 생각이 들었습니다 ..!

해당 컴포넌트는 `edit`이 아니어도 Avatar 자체에 대한 `click 이벤트`가 필요하기 때문에 client 컴포넌트로 가야 합니다. 이런 상황에서도 어떻게 분리하면 상위 부모 태그는 서버 컴포넌트로 만들 수 있겠죠..! 그렇지만 상위 태그 하나 존재하는 `Avatar`로 만든 느낌이 되는 건데.. 이렇게 해도 사실상 상위 뎁스를 제외하면 RCC 다보니 hydration이 일어나야 하위 컴포넌트가 동작하지 않나요?? 뎁스만 깊어지는 건 아닌지 라는 생각이 들었습니다.

또 하나는 지금과 다르게 Image 태그를 공용으로 사용하지 않을 거면 Avatar를 하나의 컴포넌트로 사용하는 것보다 분리해서 사용하는 측면이 더 좋지 않을까라는 생각도 들었네요 😂

Jae0o replied:
분리하는것도 좋은 생각인것 같네요 그리고 RCC의 코드 비용이 적게 들어간다면 RSC로 만들 필요는없을것같네요.

위 코드상태로는 저부분을 분리하면 RSC로 보여질 수 있을거라고 생각했는데 edit 기능외 일반 Avatar도 이벤트가 존재하면 RCC도 좋은 선택인것같아요!

Jae0o commented on src/app/_components/client/Avatar/Avatar.tsx:
```diff
@@ -0,0 +1,64 @@
+      {type === "edit" && (
+        <>
+          <span
+            className={`absolute bottom-0 right-0 ${avatarIconSize} bg-[#F1F1F1] rounded-full`}
```
[ P2 ]

bg 에 대한 color를 임시로 example color로 넣어두는건 어떨까요??

SeungHyune replied:
아하 ~ 기본 컬러는 하얀색으로 추가해 놓겠습니다 ㅎㅎ
`box-shadow`도 흰색 배경과 구분 가도록 추가해놓겠습니다! 👍

Jae0o commented on src/app/_components/client/Avatar/Avatar.tsx:
```diff
@@ -0,0 +1,64 @@
+
+  return (
+    <div
+      className={`relative ${avatarImageSize} ${(userId || type === "edit") && "cursor-pointer"}`}
```
[ P2 ]

해당 부분에서 twJoin을 통해 위 2개의 변수없이 바로 사용하는건 어떨까요?
라인단위로 구분되어지기도 하고 코드량도 많이 줄어들어서 보기편한것같아서요!

 + 아래 span 태그의 className도!..

```suggestion
      className={twJoin(
        "relative",
        AVATAR_SIZE[size],
        (userId || type === "edit") && "cursor-pointer"
      )}
```

SeungHyune replied:
`twJoin`을 많이 사용하면 생기는 단점은 없나요 ??!
일반 className을 사용했을 때와 가독성 측면을 제외하고 다른 이점은 없는지 궁금합니다 !! 😀

### twJoin 장점
- 클래스 간 충돌을 해결하지 않고도 조건부로 클래스를 결합할 수 있습니다.
- 코드의 가독성과 유지보수성이 향상됩니다.
- 사용자 정의 클래스를 만들지 않아도 됩니다.
- 번들 사이즈를 줄일 수 있습니다.

Jae0o replied:
오히려 className에 조건부를 걸고 조합하는 과정에서 발생하는 문제들을 해결하기위해 twJoin 과 twMerge 플로그인이 생겨난거라 좋은것같네요

twJoin은 현재까지 단점이나 오류는 발견한적없지만

twMerge는 가끔 사용하는 쪽에서 전달하는 추가 className으로 인해 변수적용이 안되는 버그가 있는건 확인했습니다.

하지만 twMerge가 없으면 결국 추가적으로 전달하는 className 이 미적용된다는 큰 문제가 있어서 이부분도 장점이 더 큰것같네요

Jae0o commented on src/app/_components/client/Avatar/hooks/useClickAvatar/useClickAvatar.tsx:
```diff
@@ -0,0 +1,28 @@
+const useClickAvatar = ({ type, inputRef, userId, onChangeAvatar }: useClickAvatarProps) => {
+  const router = useRouter();
+
+  const handleClickAvatar = () => {
```
[ P4 ]

어느 환경에서 hook이 호출될지 모르다보니 저는 hook 내부의 method는 useCallback이나 useRef, useMemo로 메모이제이션 해주는 편인데 어떠실까효

SeungHyune replied:
`useCallback`을 사용해서 최적화하도록 하겠습니다!! ㅎㅎ 감사합니다 👍

Jae0o reviewed (commented):
대체로 잘만들어주신것 같아요!

아까 말씀드린 부분과 동일하게 Avatar 하나의 Component에서 들어오는 User의 데이터 수에따라 자동적으로 Avatar List를 그려줄지 Avatar 를 보여줄지에 대해 기능이 추가되면 아주 베스트일것 같습니다! 

리뷰에서 이해안가는 부분있으시면 언제든 콜 주십시요!

Jaeppetto commented on src/app/_components/client/AvatarList/AvatarList.tsx:
```diff
@@ -0,0 +1,96 @@
+              className="mr-2 rounded-full overflow-hidden"
+              key={id}
+            >
+              <Avatar
```
P5;

해당 아바타 Hover 시에 닉네임 같은 걸 띄워줘도 좋을 것 같네요!

Jaeppetto commented on src/app/_components/client/AvatarList/hooks/useClickAvatarList/useClickAvatarList.tsx:
```diff
@@ -0,0 +1,21 @@
+import { useState } from "react";
+
+const useClickAvatarList = () => {
```
P5;

`AvatarList`를 클릭했을 때의 동작에 대해 고민이 많다고 하셨어서, 제 의견을 살짝 공유드려봅니다 🤗

- 현재 토글을 활성화 / 비활성화하는 경우 화면 밖에서 AvatarList가 나타나고 사라지는 것으로 보이는데요..! 아예 화면 밖으로 빼지 않고 **기존 리스트가 펼쳐지는 형식**으로 구현돼도 괜찮을 것 같다는 생각입니다 ㅎ_ㅎ
- background 뿐 아니라 useClickAway를 통해 **아예 관련 없는 부분을 클릭해도 toggle이 닫히면** 더 좋을 것 같다는 생각이 듭니다.

Songhee99 replied:
오 아예 상관없는 부분을 클릭할 때도 토글이 닫히면 더 편리할 것 같네요 ! 저도 이 아이디어에 한표입니다 ㅎㅎ

SeungHyune replied:
좋은 의견인 것 같습니다 !! 수정해 보도록 하겠습니다 ㅎㅎ 👏👍

Jaeppetto commented on src/app/_components/client/Avatar/Avatar.tsx:
```diff
@@ -0,0 +1,64 @@
+}: AvatarProps) => {
+  const { preview, inputRef, handleChangeImage } = useChangeImage({ avatarImage, onChangeAvatar });
+  const { handleClickAvatar } = useClickAvatar({ type, inputRef, userId, onChangeAvatar });
+  const avatarImageSize = AVATAR_SIZE[size];
```
P5;

오 이런식으로 프롭과 스타일 클래스를 매칭시켜줄수도 있군요! 너무 좋은 것 같습니다.

SeungHyune replied:
`tailwind`가 `w-[${data}]` 이러한 동적 데이터를 못 읽다 보니 Avatar 사이즈를 미리 정의해서 사용했습니다 !! 😊

Jaeppetto commented on src/app/_components/client/AvatarList/AvatarList.tsx:
```diff
@@ -0,0 +1,96 @@
+import Image from "next/image";
+
+interface AvatarProps {
+  id: string;
```
P5;

사소한 부분이지만 기존 Avatar.tsx에서 선언하신 프롭의 타입과 일관성을 맞추어 `userId`로 관리해도 좋을 것 같습니다!

SeungHyune replied:
넵 !! 좋은 제안이신 것 같습니다 수정하도록 하겠습니다 ! 👍

Jaeppetto reviewed (approved):
컨벤션, 컴포넌트 훅도 잘 구성된 것 같고, 상황에 따른 에러와 예외 처리도 고심하신 흔적을 엿볼 수 있었습니다. 고생많으셨습니다! 👍🏻

Songhee99 reviewed (approved):
승현님 정말 고생많으셨습니다 ㅎㅎ
코드가 꼼꼼하고 규칙도 잘 지켜진 것 같아 잘 읽혔습니다 ㅎㅎ 멋지십니다~ 👍🏻👍🏻

SeungHyune committed:
🗑️remove: AvatarList 컴포넌트 제거
- Avatar 컴포넌트 내부 컴포넌트로 통합했습니다.

SeungHyune committed:
omm################
💎feat: useToggle hook 추가
- useToggle hook은 하나의 상태 isShow를 true, false, 현재 상태의 반댓값으로 변경할 수 있는 간단한 훅입니다.

SeungHyune committed:
📝docs: userType 임시로 정의
- userResponse 값이 정해지면 Avatar props에 전달되는 user 객체 타입과 함께 변경하려고 합니다.

SeungHyune committed:
💎feat: Avatar type Edit 로직을 AvatarEdit 컴포넌트로 분리

SeungHyune committed:
💎feat: 기존 AvatarList 컴포넌트를 Avatar 컴포넌트에 포함
- 코드는 비슷하나 내부 로직을 Avatar와 통합했습니다.

SeungHyune committed:
🛠️refactor: useChangeImage, useClickAvatar hook 최적화
- useCallback을 통한 성능 최적화

SeungHyune committed:
🛠️refactor: Avatar 컴포넌트 수정
- AvatarList, Avatar를 Avatar 컴포넌트 내부에서 props 데이터에 따라 생성되도록 수정
- 기존 코드를 내부 컴포넌트로 분리

SeungHyune committed:
🛠️refactor: Avatar 컴포넌트 Storybook 수정
- 내부 props 데이터 변경
- AvatarList 컴포넌트 Avatar와 결합

Jae0o commented on src/app/_components/client/Avatar/components/AvatarList/AvatarList.tsx:
```diff
@@ -0,0 +1,76 @@
+            </div>
+          ))}
+        </div>
+        {isShow && (
+          <AnimatePresence>
+            <motion.ul
+              className="absolute left-0 top-0 z-10 flex"
+              initial={{ x: "-100%" }}
+              animate={{ x: 0 }}
+              exit={{ x: "-100%" }}
+              whileTap={{ scale: 0.95 }}
+            >
+              {users.map((user) => (
+                <li
+                  className="mr-2 overflow-hidden rounded-full"
+                  key={user.userId}
+                >
+                  <Avatar user={user} />
+                </li>
+              ))}
+            </motion.ul>
+          </AnimatePresence>
+        )}
```
[ P1 ]

해당 부분에서 Exit 동작하게 하시려면

```suggestion
        <AnimatePresence>
          {isShow && (
            <motion.ul
              className="absolute left-0 top-0 z-10 flex"
              initial={{ x: "-100%" }}
              animate={{ x: 0 }}
              exit={{ x: "-100%" }}
              whileTap={{ scale: 0.95 }}
            >
              {users.map((user) => (
                <li
                  className="mr-2 overflow-hidden rounded-full"
                  key={user.userId}
                >
                  <Avatar user={user} />
                </li>
              ))}
            </motion.ul>
          )}
        </AnimatePresence>
```

위와같이 `AnimatePresence` 태그 내부에서 상태값에 의해 언마운트 마운트가 발생해야합니다!

SeungHyune replied:
넵 !! 감사합니다 !! ㅎㅎ 👏😊
`AnimatePresence` 내부 자식이 언마운트 시에 발동되는군요 ?!

Jae0o commented on src/app/_components/client/Avatar/components/AvatarList/AvatarList.tsx:
```diff
@@ -0,0 +1,76 @@
+        className="relative"
+        ref={ref}
+      >
+        <div
+          className="inline-flex cursor-pointer items-center"
+          onClick={handleClickToggle}
+        >
+          {users.slice(0, 2).map(({ userId, userName, userImage }, index) => (
+            <div
+              key={userId}
+              className={twJoin(
+                "relative h-[3rem] w-[3rem] overflow-hidden rounded-full",
+                index === 1 && "-ml-[2.2rem] mt-[1rem]",
+                index === 0 && "z-10",
+              )}
+            >
+              <Image
+                className="object-cover"
+                src={userImage || profileImage}
+                alt={userName}
+                sizes="100%"
+                priority
+                fill
+              />
+            </div>
+          ))}
+        </div>
```
요건 개인적인 의견인데 클릭하여 프로필 리스트가 등장할때!
기존에 존재하던 프로필이미지들은 scale 0 + opacity 0 으로 숨겨지고
리스트가 들어가면 다시 등장하는 애니메이션이 추가되면어떨까요!

현재는 뒤에 가만히 남아있는게 어색한 너낌이 사아아아아알짝 들더라구욤

SeungHyune replied:
한 번 수정해 보겠습니다 !! 저도 해당 UI 및 애니메이션에 대한 고민이 많았는데 적용해 보고 말씀드리겠습니다 !
감사합니다 ㅎㅎ 👍👍

Jae0o reviewed (approved):
AnimatePresence 부분만 고쳐주시면 감사하겠습니다!

수고하셨어요!

SeungHyune committed:
🚜change: useToggle hook 내부 로직 네이밍 변경

SeungHyune committed:
🚜change: Avatar size "tiny" 추가
- 기존 small 보다 더 작은 사이즈 (30px)

SeungHyune committed:
🛠️refactor: AvatarList 컴포넌트 수정
- AvatarList 클릭 시 유저 리스트 보여주는 UI 위치 수정
- AvatarList 컴포넌트에 useClickAway hook 적용 (해당 요소를 제외한 곳을 클릭하면 전달한 콜백 함수 실행)

SeungHyune committed:
🎨style: Avatar Storybook 수정
- 타입 추가 및 설명, Viewport 변경
